### PR TITLE
Alternative model proposal: More OO model with pluggable methods

### PR DESCRIPTION
**Pull request by nscarcella**
### Background

When we started this Wollok implementation we took a rather functional approach, consisting of primarily dumb data objects manipulated by independent functions. We did so in part because it seemed to align nicely with our [design guidelines](https://github.com/uqbar-project/wollok-ts/wiki/General-Design-and-Main-Concerns#design-guidelines), but mainly because it was the less-resistance path to take with the technologies we are using.

This lead to the implementation of some micro-libraries to prevent logic repetition (such as [the tools module](https://github.com/uqbar-project/wollok-ts/blob/master/src/tools.ts) which contains common AST logic and [the operations object](https://github.com/uqbar-project/wollok-ts/blob/master/src/interpreter.ts#L273) which provides shortcuts for evaluation state manipulation), plus an ever growing number of isolated auxiliar functions and local definitions. This approach, although neat and clear, makes it hard for newcomers to identify what operations are available and how to obtain them, propitiating code repetition and steepening the entrance curve.

We also found out in time that the immutable functional approach was not fast enough for the interpreter execution and replaced it with an effect-oriented implementation instead, in which the separation of data and logic makes it harder to encapsulate and control the state manipulation (e.g. we cannot enforce a single entry point for adding instances during an evaluation, we can only suggest to always do it through the [addInstance function](https://github.com/uqbar-project/wollok-ts/blob/master/src/interpreter.ts#L293) which in the past has caused issues like number duplication).
 
Finally, we teach on TADP that is in general a better idea on hybrid Functional-OO technologies to try and find a balance between the tools of booth paradigms when available, instead of completely discarding one in favor of the other, so we thought we might try to put our money where our mouth is.

### Motivation For This Change
- Improve the discoverability of operations for all data structures.
- Better restrict operations on nodes based on the pipeline stage.
- Gain better control over state changes of mutable data.

### General Approach
The proposal consist in **adding logic to our data-objects** making them expose useful operations and improving cohesiveness in general, while **maintaining the [the language pipeline's main entry points](https://github.com/uqbar-project/wollok-ts/wiki/General-Design-and-Main-Concerns#the-language-pipeline-flow) as independent functions** so they are easily composable and rewireable.

The main setback for modelling with non-data-objects is that json is not capable of representing functions, and we want to remain able to to receive an evaluation state or AST in any stage of the pipeline. For example, we might want to support receiving an already parsed AST (maybe from Mumuki?) and continue the pipeline from there, or provide a service that applies validations to an already linked environment or even an online debugger that needs to receive an evaluation state.

For this reason we are staying away from TypeScript Classes, since the conversion from and to json is not trivial and requires a lot of boilerplate. Instead, we propose having functions that "pump-up" stupid data objects obtained from the web, injecting them with the methods needed to fulfill their expected types.

So, the general flux would be:
- Model types are defined with methods, scoped to stages just as any other field.
- We provide behavior functions (need a better name for this) for every stage, that turn data-objects into rich-objects.
- Each pipeline transition asumes it works on a rich-object and makes sure to run the right behavior function for the next stage on its output.
- At any point on the pipeline we could receive a data-object for that stage, execute the behavior function on it, and then proceed with the pipeline.
- Added methods are ignored on the `Json.stringify` method, so conversion to data-object is smooth.

### Change Summary
- Replaced tools and operations with node interface so autocomplete will offer what can be done.
- Refactored model complex types into simpler interfaces, which improves greatly the IDE performance.
- Simplified many operation signatures.
- Added new behavior functions that turn data-objects into rich-objects.
- Improved the way things are mocked on tests.
- Some logic refactor.

### Future Work
- Encapsulate effects on evaluation
- Make a deeper dive into each stage and pick useful abstractions to extract into methods.
- Maybe some of the hardest work we do during linking can be replaced by a lazy method.
- Replace the ugly cache implementation with something encapsulated on the method side.

### CODE REVIEW

**Comment by nscarcella** [comment]
@npasserini @PalumboN @fdodino I know you sitll haven't dig very deep on the codebase, so I'm not expecting you guys to do a line by line review of this PR, but if you could give your opinion on the general idea I would greatly appreciate it.

File: src/behavior.ts [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    },
+
+    transform<R extends Stage>(
+      this: Node<RawStage>,
+      tx: ((node: Node<RawStage>) => Node<R>) | { [K in Kind]?: (node: Node<R>) => Node<R> }
+    ): Node<R> {
+      const applyTransform = (value: any): any => {
+        if (typeof value === 'function') return value
+        if (isArray(value)) return value.map(applyTransform)
+        if (isNode<RawStage>(value)) return typeof tx === 'function'
```
Comment (npasserini) [inline src/behavior.ts]: Ah pero sos alto hacker.

File: src/behavior.ts [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+    },
+
+    reduce<T>(this: Node<RawStage>, tx: (acum: T, node: Node<RawStage>) => T, initial: T): T {
+      return this.children().reduce((acum, child) => child.reduce(tx, acum), tx(initial, node))
+    },
+
+  })
+
+  if (isModule(node)) assign(node, {
+    methods(this: Module<RawStage>) { return this.members.filter(is('Method')) },
```
Comment (npasserini) [inline src/behavior.ts]: Y members dónde está definido?

File: src/behavior.ts [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    descendants<T extends Node<RawStage>>(this: Node<RawStage>, filter?: (node: Node<RawStage>) => node is T): List<T> {
+      const directDescendants = this.children<Node<RawStage>>()
+      const indirectDescendants = flatMap<Node<RawStage>>(child => child.descendants(filter))(directDescendants)
+      const descendants = [...directDescendants, ...indirectDescendants]
+      return filter ? descendants.filter(filter) : descendants as any
+    },
+
+    transform<R extends Stage>(
+      this: Node<RawStage>,
+      tx: ((node: Node<RawStage>) => Node<R>) | { [K in Kind]?: (node: Node<R>) => Node<R> }
```
Comment (npasserini) [inline src/behavior.ts]: El penúltimo `Node<R>` no debería ser `Node<RawStage>`?

File: src/behavior.ts [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+// ══════════════════════════════════════════════════════════════════════════════════════════════════════════════════
+
+export const Linked = (environmentData: Partial<Environment>) => {
+  const environment: Environment = assign(Filled(environmentData as any), {
+
+    getNodeById<T extends Node<LinkedStage>>(this: Environment, id: Id): T {
+      return getOrUpdate(NODE_CACHE, id)(() => {
+        const search = (obj: any): Node<LinkedStage> | undefined => {
+          if (isArray(obj)) {
+            for (const value of obj) {
```
Comment (npasserini) [inline src/behavior.ts]: Qué representa `obj` acá? Por qué devuelve el primero que encuentra?

File: src/behavior.ts [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+const { values, assign, keys } = Object
+
+// TODO: Test all behaviors
+
+// ══════════════════════════════════════════════════════════════════════════════════════════════════════════════════
+// RAW
+// ══════════════════════════════════════════════════════════════════════════════════════════════════════════════════
+
+export const Raw = <N extends Node<RawStage>>(obj: Partial<N>): N => {
+  const node = { ...obj } as N
```
[the remark this answers is not in the document]
Comment (nscarcella) [reply src/behavior.ts]: Vos decís el casteo?

File: src/behavior.ts [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+  assign(node, {
+
+    children<T extends Node<RawStage>>(this: Node<RawStage>): List<T> {
```
[the remark this answers is not in the document]
Comment (nscarcella) [reply src/behavior.ts]: Chiche. Esa es la notación para marcarle a una función el tipo de this.

File: src/behavior.ts [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+      const extractChildren = (owner: any): List<T> => {
+        if (isNode(owner)) return [owner as T]
+        if (isArray(owner)) return flatMap(extractChildren)(owner)
+        if (owner instanceof Object) return flatMap(extractChildren)(values(owner))
+        return []
+      }
+
+      const extractedChildren = flatMap(extractChildren)(values(this))
+      extractedChildren.forEach(child => child.id && update(PARENT_CACHE, child.id!, this.id))
```
[the remark this answers is not in the document]
Comment (nscarcella) [reply src/behavior.ts]: Aich... Esa caché es una mierda que tuve que poner para mejorar la performance. Básicamente como el AST es un árbol sin ciclos recorrerlo hacia arriba (por ejemplo, buscar el padre de un nodo) lleva mucho tiempo. Entonces lo cacheamos. Mi plan es pensar si tener métodos me ayuda a volar esa cache.

File: src/behavior.ts [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+      return extractedChildren
+    },
+
+    descendants<T extends Node<RawStage>>(this: Node<RawStage>, filter?: (node: Node<RawStage>) => node is T): List<T> {
+      const directDescendants = this.children<Node<RawStage>>()
+      const indirectDescendants = flatMap<Node<RawStage>>(child => child.descendants(filter))(directDescendants)
+      const descendants = [...directDescendants, ...indirectDescendants]
+      return filter ? descendants.filter(filter) : descendants as any
```
[the remark this answers is not in the document]
Comment (nscarcella) [reply src/behavior.ts]: Sí, puede ser. Capaz se podría implementar con el método reduce del nodo o reemplazar por una function* que yieldee de a uno. Habría que medir si mejora los tiempos/memoria

File: src/behavior.ts (hunk 3 above)
Comment (nscarcella) [reply src/behavior.ts]: A la mierda, estás prestando atención... Sí, debería. En un rato lo arreglo.

File: src/behavior.ts (hunk 1 above)
Comment (nscarcella) [reply src/behavior.ts]: Este es el método más feo del proyecto.

File: src/behavior.ts (hunk 2 above)
Comment (nscarcella) [reply src/behavior.ts]: Cada Module tiene su propia definición. Acá el TS se aviva y me deja usarlo después del smart cast

File: src/behavior.ts (hunk 4 above)
Comment (nscarcella) [reply src/behavior.ts]: Esto se corre recursivamente sobre un nodo del AST, entonces obj es un nodo o un campo de un nodo. Retorna el primero porque es una búsqueda por Id y los id's son únicos

**Pull request merged**